### DISCUSSION TITLE
`jsonable_encoder` on SQLAlchemy mapper object causing maximum recursion error

### DISCUSSION THREAD

**Discussion by xuzuoyang** [fastapi/fastapi#9026] [Questions]
### Example

```Python

# First create two mapper classes based on SQLALchemy
class User(Base):
    ...
    profile = relationship('Profile', backref='user')

class Profile(Base):
    pass

# This way we can use either user.profile / profile.user to reference each other

profile.user  # load the attrs for profile first
jsonable_encoder(user)  # and here it will cause maximum recursion depth exceeded error
```

### Description

- To reproduce the error `profile.user` has to be called before `jsonable_encoder`, since there's a `lazy load` feature provided by SQLAlchemy
- I looked up the source code and think the root cause is the use of `vars()` since there's not a specific branch for SQLAlchemy mapper object:

```Python
    ...
    try:
        data = dict(obj)
    except Exception as e:
        errors.append(e)
        try:
            data = vars(obj)
        except Exception as e:
            errors.append(e)
            raise ValueError(errors)
    ...
```

### Environment

- OS: macOS
- FastAPI Version: 0.63.0
- Python Version: 3.8.2

**Comment by ycd** [comment]
I think you can create a custom encoder to handle this.

**Comment by Kludex** [comment]
Can you show us your endpoint function and the pydantic output model definition?

**Comment by xuzuoyang** [comment]
> I think you can create a custom encoder to handle this.

Thanks, I see custom encoder can be used as a workaround, while a general solution is still expected.

**Comment by xuzuoyang** [comment]
> Can you show us your endpoint function and the pydantic output model definition?

Actually the example code is not what I used in the real work and I encountered this issue in one of my test cases. Are you saying this may not be a problem for proper designed api functions?

**Comment by bhzunami** [comment]
I think this issue is resolved in the version 0.65.1. I had the same issue and resolved the error with upgrading to FastAPI version 0.65.1 and sqlalchemy to version 1.4.15. Not sure if the upgrade to sqlalchemy 1.4.15 was really necessary.

**Comment by xuzuoyang** [comment]
> I think this issue is resolved in the version 0.65.1. I had the same issue and resolved the error with upgrading to FastAPI version 0.65.1 and sqlalchemy to version 1.4.15. Not sure if the upgrade to sqlalchemy 1.4.15 was really necessary.

Thanks for the update, I'll try it out today.

**Comment by xuzuoyang** [comment]
@bhzunami I tested under 0.65.2, `RecursionError` still stands. Not sure if we have encountered the same issue here, do you mind sharing your case?

**Comment by danielstopp** [comment]
I'm leaving this comment here for clarity, but this is NEITHER a FastAPI issue NOR an SQLAlchemy or uvicorn one. I was simply being dumb (see my follow-up comment).

Better check whether you're making the same mistake, before posting here ;)

---

I have th same problem with infinite recursion. I'm using FastAPI, SQLAlchemy and uvicorn with Python 3.9.5 inside inside a Docker Compose setup, here's the error log:

```
web_1       | INFO:     172.20.0.1:46896 - "POST /users/ HTTP/1.1" 500 Internal Server Error
web_1       | ERROR:    Exception in ASGI application
web_1       | Traceback (most recent call last):
web_1       |   File "/usr/local/lib/python3.9/site-packages/uvicorn/protocols/http/httptools_impl.py", line 371, in run_asgi
web_1       |     result = await app(self.scope, self.receive, self.send)
web_1       |   File "/usr/local/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 59, in __call__
web_1       |     return await self.app(scope, receive, send)
web_1       |   File "/usr/local/lib/python3.9/site-packages/fastapi/applications.py", line 199, in __call__
web_1       |     await super().__call__(scope, receive, send)
web_1       |   File "/usr/local/lib/python3.9/site-packages/starlette/applications.py", line 112, in __call__
web_1       |     await self.middleware_stack(scope, receive, send)
web_1       |   File "/usr/local/lib/python3.9/site-packages/starlette/middleware/errors.py", line 181, in __call__
web_1       |     raise exc from None
web_1       |   File "/usr/local/lib/python3.9/site-packages/starlette/middleware/errors.py", line 159, in __call__
web_1       |     await self.app(scope, receive, _send)
web_1       |   File "/usr/local/lib/python3.9/site-packages/starlette/exceptions.py", line 82, in __call__
web_1       |     raise exc from None
web_1       |   File "/usr/local/lib/python3.9/site-packages/starlette/exceptions.py", line 71, in __call__
web_1       |     await self.app(scope, receive, sender)
web_1       |   File "/usr/local/lib/python3.9/site-packages/starlette/routing.py", line 580, in __call__
web_1       |     await route.handle(scope, receive, send)
web_1       |   File "/usr/local/lib/python3.9/site-packages/starlette/routing.py", line 241, in handle
web_1       |     await self.app(scope, receive, send)
web_1       |   File "/usr/local/lib/python3.9/site-packages/starlette/routing.py", line 52, in app
web_1       |     response = await func(request)
web_1       |   File "/usr/local/lib/python3.9/site-packages/fastapi/routing.py", line 214, in app
web_1       |     raw_response = await run_endpoint_function(
web_1       |   File "/usr/local/lib/python3.9/site-packages/fastapi/routing.py", line 151, in run_endpoint_function
web_1       |     return await run_in_threadpool(dependant.call, **values)
web_1       |   File "/usr/local/lib/python3.9/site-packages/starlette/concurrency.py", line 40, in run_in_threadpool
web_1       |     return await loop.run_in_executor(None, func, *args)
web_1       |   File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 52, in run
web_1       |     result = self.fn(*self.args, **self.kwargs)
web_1       |   File "/usr/src/app/./app/entrypoints/fast_api.py", line 102, in create_user
web_1       |     new_user = create_user(db=db, user=user)
web_1       |   File "/usr/src/app/./app/entrypoints/fast_api.py", line 102, in create_user
web_1       |     new_user = create_user(db=db, user=user)
web_1       |   File "/usr/src/app/./app/entrypoints/fast_api.py", line 102, in create_user
web_1       |     new_user = create_user(db=db, user=user)
web_1       |   [Previous line repeated 990 more times]
web_1       | RecursionError: maximum recursion depth exceeded
```

I'm using the following libs/packages:

```
uvicorn[standard]==0.14.0
fastapi==0.65.2
email-validator==1.1.3
passlib==1.7.4
psycopg2-binary==2.9.1
SQLAlchemy==1.4.20
pytest==6.2.4
```

Any clues, so far, what could be causing this? Is this really a FastAPI issue or could this be caused by uvicorn or SQLAlchemy as well?

**Comment by danielstopp** [comment]
I resolved the issue and it was purely my own fault.

I caused the infinite recursion myself by putting everything in a single file for testing out FastAPI in combination with SQLAlchemy and Pydantic, by failing to see that I gave the CRUD functon and the route function the same name, so the route function effectively shadowed the CRUD function, leaving to an infinite loop.

Let that be a warning to you, lest you have to look like an idiot :dagger: 

Sorry for the fuss.

**Comment by bhzunami** [comment]
@xuzuoyang I have my code here: https://github.com/bhzunami/bedello_backend
And here https://github.com/bhzunami/bedello_backend/blob/master/app/controller/base_controller.py#L40 I use the jsonable_encoder function (Most of the code is copy paste from the fastapi postgresql repository)

**Comment by xuzuoyang** [comment]
> I resolved the issue and it was purely my own fault.
> 
> I caused the infinite recursion myself by putting everything in a single file for testing out FastAPI in combination with SQLAlchemy and Pydantic, by failing to see that I gave the CRUD functon and the route function the same name, so the route function effectively shadowed the CRUD function, leaving to an infinite loop.
> 
> Let that be a warning to you, lest you have to look like an idiot 
> 
> Sorry for the fuss.

Thank you for the warning, I'll keep that in mind. In my case, I've checked the log with some breakpoint, pretty sure it has sth to do with the source code.

**Comment by xuzuoyang** [comment]
> @xuzuoyang I have my code here: https://github.com/bhzunami/bedello_backend
> And here https://github.com/bhzunami/bedello_backend/blob/master/app/controller/base_controller.py#L40 I use the jsonable_encoder function (Most of the code is copy paste from the fastapi postgresql repository)

Thanks, I'll look it up.

**Comment by xuzuoyang** [comment]
@bhzunami Seems like the only model you use in the current api `Category`, which has a one-to-many relationship with `Product`, won't cause the issue. On the other hand, `Customer` and `User` may get to be a real case, since they can be referenced by each other recursively, such as `c.u.c`. Anyway, thanks for the code, I'll setup a case example myself and update for this issue.

**Comment by xuanlinh91** [comment]
I still have this problem in 2023, Can someone tell me how to fix this?

**Comment by todofixthis** [comment]
Here's a partial solution to the problem.  Basically it wraps `jsonable_encoder` and filters out any 'child' relationships from the object before passing it along to `jsonable_encoder`.

**Note:** Doesn't account for one-to-one and many-to-many relationships, mostly because I haven't yet worked out how best to do this 😇

```python
from fastapi.encoders import jsonable_encoder
from sqlalchemy import inspect
from sqlalchemy.orm import ONETOMANY
from typing import Sequence

def model_encoder(model: Any, **kwargs) -> dict:
    """
    Safely converts a model instance into a dict, so it can be JSON-encoded.

    :param model: Model instance, but for convenience will accept any value.
    :param kwargs: Additional kwargs to pass to :py:func:`jsonable_encoder`.

    :see: https://github.com/tiangolo/fastapi/discussions/9026
    :see: https://github.com/sqlalchemy/sqlalchemy/issues/9785
    """
    if isinstance(model, Base):
        mapper = inspect(model).mapper

        # Add non-relationship values.
        cleaned = {key: getattr(model, key) for key in mapper.columns.keys()}

        # Add "parent" (one-to-many) relations.
        cleaned.update(
            {
                key: [model_encoder(value) for value in getattr(model, key)]
                for key, relationship in mapper.relationships.items()
                if relationship.direction == ONETOMANY
            }
        )
    elif isinstance(model, Sequence):
        cleaned = [model_encoder(value) for value in model]
    else:
        cleaned = model

    return jsonable_encoder(cleaned, **kwargs)
```

**Comment by YuriiMotov** [comment]
Please, provide MRE
